pkg/storage/replica.go

```diff
@@ -4147,50 +4195,50 @@ func (r *Replica) executeBatch(
 		if pErr != nil {
 			switch tErr := pErr.GetDetail().(type) {
 			case *roachpb.WriteTooOldError:
-				// WriteTooOldErrors may be the product of raft replays. If
-				// timestamp of the request matches exactly with the existing
-				// value, maybe propagate the WriteTooOldError to let client
-				// retry at a higher timestamp. Keep in mind that this replay
-				// protection is best effort only. If replays come out of
-				// order, we'd expect them to succeed as the timestamps which
-				// would match on a successive replay won't match if the replay
-				// is delivered only after another raft command has been applied
-				// to the same key.
-				if ba.Timestamp.Next().Equal(tErr.ActualTimestamp) {
-					// If in a txn, propagate WriteTooOldError immediately. In
-					// a txn, intents from earlier commands in the same batch
-					// won't return a WriteTooOldError.
-					if ba.Txn != nil {
-						return nil, result, pErr
-					}
-					// If not in a txn, need to make sure we don't propagate the
-					// error unless there are no earlier commands in the batch
-					// which might have written the same key.
-					var overlap bool
+				// We got a WriteTooOldError. In case this is a transactional request,
+				// we want to run the other commands in the batch and let them lay
+				// intents so that other concurrent overlapping transactions are forced
+				// through intent resolution and the chances of this batch succeeding
+				// when it will be retried are increased.
+				// Except in case the ba.WillNotBeRetried bit is set we don't want the
+				// the intents to be laid - it would be futile.
+				if ba.WillNotBeRetried {
+					nonTxnOverlappingBatch := false
+					// If this is a non-transactional batch, then the WriteTooOldError
+					// can be caused by a previous overlapping command in the batch (in
+					// transactional batches, running into your own intent doesn't cause a
+					// WriteTooOldError). If this is the case, we want to swallow the
+					// error and continue execution.
 					if ba.Txn == nil {
 						for _, union := range ba.Requests[:index] {
 							if union.GetInner().Header().Overlaps(args.Header()) {
-								overlap = true
+								nonTxnOverlappingBatch = true
 								break
 							}
 						}
 					}
-					if !overlap {
+					if !nonTxnOverlappingBatch {
 						return nil, result, pErr
 					}
 				}
+
 				// On WriteTooOldError, we've written a new value or an intent
 				// at a too-high timestamp and we must forward the batch txn or
 				// timestamp as appropriate so that it's returned.
 				if ba.Txn != nil {
 					ba.Txn.Timestamp.Forward(tErr.ActualTimestamp)
 					ba.Txn.WriteTooOld = true
 				} else {
+					// Note that, when the batch appears here to be non-transactional but in
```

Comment:
s/Note that,/Note:/
s/appears here to be/appears to be/
/pushing the timestamp here/pushing the timestamp/